### PR TITLE
[Feature-2154][api] Workflow version control

### PR DESCRIPTION
**Pull request by yangyichao-mango**
## What is the purpose of the pull request

*Implement workflow version control (#2154 ) (#3046 )*

## Brief change log

  - *Add the process definition when create or update the process definition in `ProcessDefinitionController`, `ProcessDefinitionService`, and update the process instance in `ProcessInstanceService`*
  - *Add new table `t_ds_process_definition_version` to database, and add the service `ProcessDefinitionVersionService`, dao `ProcessDefinitionVersionMapper.java`, mapper `ProcessDefinitionVersionMapper.xml`, entity `ProcessDefinitionVersion`*
  - *Add some return info in `Status`*
  - *Add the `version.vue` in front-end ui and use it in `list.vue`, `dag.vue`*

## Verify this pull request

  - *Add test cases to cover this unit cases for this feature*
  - *Package and test feature and `sql` runner in linux env locally*

### CODE REVIEW

**Comment by yangyichao-mango** [comment]
image
image
image
image

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/ProcessDefinitionVersionService.java [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+     */
+    public long queryMaxVersionByProcessDefinitionId(int processDefinitionId) {
+        // TODO Is it necessary to add distributed locks?
+        Long maxVersion =  processDefinitionVersionMapper.queryMaxVersionByProcessDefinitionId(processDefinitionId);
+        if (Objects.isNull(maxVersion)) {
+            return 0L;
+        } else {
+            return maxVersion;
+        }
+    }
```
Comment (yangyichao-mango) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/ProcessDefinitionVersionService.java:97]: Hi @dailidong , Is it necessary to add distributed locks?

**Comment by yangyichao-mango** [comment]
Hi @dailidong .
When I want to add the unit test case in `ProcessDefinitionControllerTest`, I found the `ProcessDefinitionControllerTest` is ran with the `MockitoJUnitRunner` but not the `SpringRunner` like the other controller test cases. I checked the history log and found @zixi0825 changed it from `SpringRunner` to `MockitoJUnitRunner` in `ProcessDefinitionControllerTest`.

Hi @zixi0825 , can you help me to understand why changed `SpringRunner` to `MockitoJUnitRunner`, Thx a lot.

**Comment by zixi0825** [comment]
> Hi @dailidong .
> When I want to add the unit test case in `ProcessDefinitionControllerTest`, I found the `ProcessDefinitionControllerTest` is ran with the `MockitoJUnitRunner` but not the `SpringRunner` like the other controller test cases. I checked the history log and found @zixi0825 changed it from `SpringRunner` to `MockitoJUnitRunner` in `ProcessDefinitionControllerTest`.
> 
> Hi @zixi0825 , can you help me to understand why changed `SpringRunner` to `MockitoJUnitRunner`, Thx a lot.

MockitoJUnitRunner is used to simulate spring injection objects. 
like this :  Mockito.when(projectMapper.queryByName(projectName)).thenReturn(getProject(projectName));

**Comment by yangyichao-mango** [comment]
> > Hi @dailidong .
> > When I want to add the unit test case in `ProcessDefinitionControllerTest`, I found the `ProcessDefinitionControllerTest` is ran with the `MockitoJUnitRunner` but not the `SpringRunner` like the other controller test cases. I checked the history log and found @zixi0825 changed it from `SpringRunner` to `MockitoJUnitRunner` in `ProcessDefinitionControllerTest`.
> > Hi @zixi0825 , can you help me to understand why changed `SpringRunner` to `MockitoJUnitRunner`, Thx a lot.
> 
> MockitoJUnitRunner is used to simulate spring injection objects.
> like this : Mockito.when(projectMapper.queryByName(projectName)).thenReturn(getProject(projectName));

Hi @zixi0825 , thx a lot for your answer.
What I actually want to know is why we need to change the `ProcessDefinitionControllerTest` test component from `SpringRunner` to `MockitoJUnitRunner`. Is it necessary for us to keep all the ApiTestCase consistent(Replace to MockitoJUnitRunner or SpringRunner)?

非常感谢你的回答。
但是我想了解的是之前是有什么场景需要我们把`ProcessDefinitionControllerTest` 的测试组件从`SpringRunner` 换为`MockitoJUnitRunner` 吗？
我们是否有必要将api层所有的测试框架统一(全部换为MockitoJUnitRunner或者SpringRunner)？

**Comment by yangyichao-mango** [comment]
Add the export or import workflow version function.

**Comment by yangyichao-mango** [comment]
> Add the export or import workflow version function.

**Comment by zixi0825** [comment]
> > > Hi @dailidong .
> > > When I want to add the unit test case in `ProcessDefinitionControllerTest`, I found the `ProcessDefinitionControllerTest` is ran with the `MockitoJUnitRunner` but not the `SpringRunner` like the other controller test cases. I checked the history log and found @zixi0825 changed it from `SpringRunner` to `MockitoJUnitRunner` in `ProcessDefinitionControllerTest`.
> > > Hi @zixi0825 , can you help me to understand why changed `SpringRunner` to `MockitoJUnitRunner`, Thx a lot.
> > 
> > 
> > MockitoJUnitRunner is used to simulate spring injection objects.
> > like this : Mockito.when(projectMapper.queryByName(projectName)).thenReturn(getProject(projectName));
> 
> Hi @zixi0825 , thx a lot for your answer.
> What I actually want to know is why we need to change the `ProcessDefinitionControllerTest` test component from `SpringRunner` to `MockitoJUnitRunner`. Is it necessary for us to keep all the ApiTestCase consistent(Replace to MockitoJUnitRunner or SpringRunner)?
> 
> 非常感谢你的回答。
> 但是我想了解的是之前是有什么场景需要我们把`ProcessDefinitionControllerTest` 的测试组件从`SpringRunner` 换为`MockitoJUnitRunner` 吗？
> 我们是否有必要将api层所有的测试框架统一(全部换为MockitoJUnitRunner或者SpringRunner)？

1.  SpringRunner has no way to simulate object dependency injection
2. Whether it is unified depends on whether  MockitoJUnitRunner or SpringRunner can meet all needs

————————
1. SpringRunner没办法模拟对象的依赖注入
2. 是否全部替换MockitoJUnitRunner或者SpringRunner取决它们其中一个是否能满足所有需求，这个需要研究一下，不然会做无用功

**Comment by yangyichao-mango** [comment]
Thx a lot for your answer.
I will investigate this case.

**Comment by yangyichao-mango** [comment]
> > > > Hi @dailidong .
> > > > When I want to add the unit test case in `ProcessDefinitionControllerTest`, I found the `ProcessDefinitionControllerTest` is ran with the `MockitoJUnitRunner` but not the `SpringRunner` like the other controller test cases. I checked the history log and found @zixi0825 changed it from `SpringRunner` to `MockitoJUnitRunner` in `ProcessDefinitionControllerTest`.
> > > > Hi @zixi0825 , can you help me to understand why changed `SpringRunner` to `MockitoJUnitRunner`, Thx a lot.
> > > 
> > > 
> > > MockitoJUnitRunner is used to simulate spring injection objects.
> > > like this : Mockito.when(projectMapper.queryByName(projectName)).thenReturn(getProject(projectName));
> > 
> > 
> > Hi @zixi0825 , thx a lot for your answer.
> > What I actually want to know is why we need to change the `ProcessDefinitionControllerTest` test component from `SpringRunner` to `MockitoJUnitRunner`. Is it necessary for us to keep all the ApiTestCase consistent(Replace to MockitoJUnitRunner or SpringRunner)?
> > 非常感谢你的回答。
> > 但是我想了解的是之前是有什么场景需要我们把`ProcessDefinitionControllerTest` 的测试组件从`SpringRunner` 换为`MockitoJUnitRunner` 吗？
> > 我们是否有必要将api层所有的测试框架统一(全部换为MockitoJUnitRunner或者SpringRunner)？
> 
> 1. SpringRunner has no way to simulate object dependency injection
> 2. Whether it is unified depends on whether  MockitoJUnitRunner or SpringRunner can meet all needs
> 
> ————————
> 
> 1. SpringRunner没办法模拟对象的依赖注入
> 2. 是否全部替换MockitoJUnitRunner或者SpringRunner取决它们其中一个是否能满足所有需求，这个需要研究一下，不然会做无用功

Thx a lot for your answer.
I will investigate this case.

**Comment by lgcareer** [comment]
@yangyichao-mango Hi,So sorry for response so late,Because the process definition json includes datasource and resources,I think it's better to design the common version(process definition,datasource,resource).

**Comment by yangyichao-mango** [comment]
> @yangyichao-mango Hi,So sorry for response so late,Because the process definition json includes datasource and resources,I think it's better to design the common version(process definition,datasource,resource).

This is just the first version of `version control`, just include the `process definition version control`, and I will discuss about datasource and resource in next version.

**Review by qiaozhanwei — APPROVED** [review]
+1

**Comment by lgcareer** [comment]
> > @yangyichao-mango Hi,So sorry for response so late,Because the process definition json includes datasource and resources,I think it's better to design the common version(process definition,datasource,resource).
> 
> This is just the first version of `version control`, just include the `process definition version control`, and I will discuss about datasource and resource in next version.

OK,the next verson can discuss on the datasource and resource.

**Comment by geosmart** [comment]
when does this feature merge to 1.3.2-release?

**Comment by tonycnlin** [comment]
Is this feature merge to 1.3.6 or not?
I didn't see this function on the page.

**Pull request merged**